### PR TITLE
[BE-4022] Support for TZ-Aware Type in Calcite

### PR DESCRIPTION
**Pull request by njriasan**
Adds support for a new type to handle TZ-Aware Timestamps. Is used to support proper typing information in BodoSQL.

Here is some context on how Calcite handles types: https://bodo.atlassian.net/wiki/spaces/BodoSQL/pages/1226309741/Sql+Types

I've verified in BodoSQL that this appears to be correct. Once I have the BodoSQL side fully implemented I will write a confluence doc on this type.

### CODE REVIEW

**Comment by njriasan** [comment]
@allai5 The failures are known issues which are fine since we don't just support all java versions. We need fix/restrict this CI in the future.

**Comment by allai5** [comment]
> https://bodo.atlassian.net/wiki/spaces/BodoSQL/pages/1226309741/Sql+Types



> @allai5 The failures are known issues which are fine since we don't just support all java versions. We need fix/restrict this CI in the future.

I've seen failures on JDK11 and OpenJ9 8 before, but not JDK 8 Druid Tests. And looking at the past commits to `bodo-calcite-1.30.0-dev` that is the case for those as well. Can you just confirm that the JDK 8 Druid Tests failures are as expected?

**Review by allai5 — COMMENTED** [review]
This mostly looks good to me, I can't see anything out of the ordinary. 

My only comments are that I'm concerned about the Java 8 Druid tests failing and I want to see a BodoSQL test to verify this functionality, but it seems like you're already working on that.

**Comment by njriasan** [comment]
> This mostly looks good to me, I can't see anything out of the ordinary.
> 
> My only comments are that I'm concerned about the Java 8 Druid tests failing and I want to see a BodoSQL test to verify this functionality, but it seems like you're already working on that.

That issue was because I hadn't merged `bodo-calcite-1.30.0-dev` back into this branch. Its fixed now.

**Comment by njriasan** [comment]
I'm going to wait to merge this PR until I have a PR passing azure nightly on my testing. It shouldn't break anything, but it never hurts to be sure.

File: core/src/main/java/org/apache/calcite/sql/type/BodoTZInfo.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  /**
+   * Get the zone information as it will be passed to Python.
+   *
+   * @return A string of the generated Python code for
+   * the zone.
+   */
+  public String getPyZone() {
+    if (pyType.equals("str")) {
+      // Append single quotes for Python strings.
+      return String.format(Locale.ROOT, "'%s'", zone);
```
Comment (keatoooon) [inline core/src/main/java/org/apache/calcite/sql/type/BodoTZInfo.java:51]: The Locale isn't doing anything in this case, right? You're just making the string quoted

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeFactoryImpl.java [hunk 2]
```diff
@@ -126,6 +126,17 @@ public SqlTypeFactoryImpl(RelDataTypeSystem typeSystem) {
...
 
+  /**
+   * Creates a TZAware SQL type. This is a Timestamp with particular Timezone information.
+   *
+   * @param tzInfo    The timezone information
+   * @return canonical TZAware SQL type descriptor
+   */
+  @Override public RelDataType createTZAwareSqlType(BodoTZInfo tzInfo) {
+    RelDataType newType = new TZAwareSqlType(typeSystem, tzInfo, false);
+    return canonize(newType);
```
Comment (keatoooon) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeFactoryImpl.java:137]: I assume canonize just registers the type with the type factory if it hasn't been already? or something similar?

**Review by keatoooon — COMMENTED** [review]
I had a few questions to make sure I understood, but overall LGTM. I would maybe add this type to the Timestamp family similar to how TIMESTAMP_WITH_TIMEZONE is (core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java, line 116), since we'll need that in order for this type to work with most of the exiting timestamp functions that we've implemented/that exist in Calcite

File: core/src/main/java/org/apache/calcite/sql/type/BodoTZInfo.java (hunk 1 above)
Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql/type/BodoTZInfo.java:51]: Locale encodes it in the correct region specific dialect: https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html.

This is required by the calcite style checker.

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeFactoryImpl.java (hunk 2 above)
Comment (njriasan) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeFactoryImpl.java:137]: `canonize` basically ensures there is a singleton for each parameterized type. It causes a type to be registered in a cache and reuse the same object.

**Comment by njriasan** [comment]
> I had a few questions to make sure I understood, but overall LGTM. I would maybe add this type to the Timestamp family similar to how TIMESTAMP_WITH_TIMEZONE is (core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java, line 116), since we'll need that in order for this type to work with most of the exiting timestamp functions that we've implemented/that exist in Calcite

I'm happy to extend the family when it proves necessary. So far I've been matching on datetime and that seems to work. I have some concern about the existing timestamp functions "working"  since we need to ensure we actually handle timezones properly, at least in the return types. I have a couple working examples for this and so I would actually probably be happy with failures here, but I'll see what happens after running nightly. It may be necessary for the operands.

**Comment by njriasan** [comment]
> > I had a few questions to make sure I understood, but overall LGTM. I would maybe add this type to the Timestamp family similar to how TIMESTAMP_WITH_TIMEZONE is (core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java, line 116), since we'll need that in order for this type to work with most of the exiting timestamp functions that we've implemented/that exist in Calcite
> 
> I'm happy to extend the family when it proves necessary. So far I've been matching on datetime and that seems to work. I have some concern about the existing timestamp functions "working" since we need to ensure we actually handle timezones properly, at least in the return types. I have a couple working examples for this and so I would actually probably be happy with failures here, but I'll see what happens after running nightly. It may be necessary for the operands.

Alternatively we may actually want to write our own families to avoid breaking existing code.

**Review by keatoooon — APPROVED** [review]
My questions are answered, LGTM

**Pull request merged**